### ISSUE TITLE
Would you accept a pull request for gRPC-webpieces-http client and gRPC-webpieces-server?

### ISSUE THREAD

**Issue by deanorderly** [grpc/grpc-java#6984]
Looking into doing work for gRPC generation for our microservices and webpieces is a BUNCH of webpieces although with a backpressure feature wired through 'every' layer even in http1.1.  I know http2 'has' backpressure though in webpieces, the sockets get turned off on a backpressure event due to controllers not keeping up leaving CPU to concentrate on finishing existing requests before turning backpressure on.  Instead of netty, webpieces has it's own nio layer(a library nio layer not a framework like netty library so easier to compose with).

thoughts on a pull request in the future?  It may take our organization some time to create the implementation, so we would fork the repo and hopefully be back in a couple/few months to do a pull request.

**Comment by voidzcy** [comment]
We are not quite familiar with webpieces you mentioned above. It sounds like some special implementation on the transport layer. Could you provide more information about what it is and contexts about your use case? Then we may be able to decide how we should proceed it.

**Comment by deanorderly** [comment]
@voidzcy Sure, there is much to explain though as each piece really needs an explanation.  Let me start with it's a spin-off of the java 'playframework' project.  This is very important as it has a library for hot-recompile(via eclipse) in it's devserver(not production) so you can rapid develop without restarting the java server.  **We hope to rapid develop with no server restarts with gRPC.**

Here is a picture before I go into the pieces of webpieces which does not show the http1.1 client nor the http2 client...

https://staging-admin-pkyroyoegq-uw.a.run.app/@documentation/design#webpiecesDesign

I want to do both gRPC client and gRPC server,  BUT let's just talk server first...

Looking at the design picture, you can see that I could follow the Netty provider and wrap 'http-frontend' with gRPC Server **but then we lose hot-recompile and no server restarts :( :(.**  

What I 'may' try to do (but very worried about apis in gRPC changing and breaking us) is that I would like to try to have an api into gRPC so instead of plugging into java-gRPC 'Server' or gRPC 'ChannelProvider' for clients..... instead, I am wondering if I can reverse it so there is some io.grpc.api that I can call as http2 objects come in and I get back grpc protobuf objects to feed back into the api which will give me the bytes to send back.

Let me give you more detail on the webpieces side of routing.  We would like to bind these routes to grpc...
https://host/grpc/{grpcService}/{grpcMethod}
https://host/json/{grpcService}/{grpcMethod}

leaving other routes for webpages, etc. whatever

A controller (in a soon to be released version) can have this method and this wasn't just for grpc

@Override
public CompletableFuture<StreamWriter> incomingRequest(String grpcService, String grpcMethod, Http2Request headers, ResponseStream stream);

At this point, when received, the StreamWriter that is returned can be written to with data as well.  This is the webpieces wire-in point.  EVERYTHING below this method is hot-recompiled so changes to protobuf files etc. can be done WITH NO RESTART to the server.  

SIDE NOTE: We return CompletableFuture<...> instead of StreamWriter as this is how backpressure works in webpieces.  If the controllers do not return enough resolved futures, backpressure starts to kick in(even on http1.1).

thanks,
Dean

**Comment by deanorderly** [comment]
I am not sure grpc can handle the https://host/json/{grpcService}/{grpcMethod} yet though as I do ask that side question here https://host/json/{grpcService}/{grpcMethod}

perhaps gRPC needs a separate feature request to generate simpler request and response POJO's so that when gRPC stubs are generated, the same service and objects can be used by an incoming json and incoming protobuf bytes???  (Currently, it seems gRPC makes it a bit tough to do both on the same service from the blogs I have read and seems tons of people are searching on that as I looked around).

**Comment by deanorderly** [comment]
hmmm, thinking more, can't gRPC simply expose that as the API!!!  ie. if this was the gRPC api, that would be BOMB

public CompletableFuture<...StreamWriter..> incomingRequest(String grpcService, String grpcMethod, Http2Request headers, ResponseStream stream);

I think that is all the info that is needed.  the rest can be left to the platform to do TLS/SSL.  gRPC should not care about that stuff, or should it?

**Comment by deanorderly** [comment]
@voidzcy I realize you are very busy but any thoughts?

**Comment by ejona86** [comment]
@deanorderly, it still isn't clear what the webpieces project is. You're talking about what it is achieving. Could you link to a public website or git repository?

Right now gRPC generated-code-based services implement [`BindableService`](https://grpc.github.io/grpc-java/javadoc/io/grpc/BindableService.html). That wraps up the user's code and produces a [`ServerServiceDefintion`](https://grpc.github.io/grpc-java/javadoc/io/grpc/ServerServiceDefinition.html), of which the [`ServerMethodDefinition`s](https://grpc.github.io/grpc-java/javadoc/io/grpc/ServerMethodDefinition.html) are the important part.

The `MethodDescriptor` has the method name and details how to serialize/deserialize the messages. And the `ServerCallHandler` is what is called by gRPC:

```java
ServerCall.Listener<RequestT> startCall(ServerCall<RequestT,ResponseT> call, Metadata headers)
```

I'd suggest making an adapter to convert from the `incomingRequest()` webpieces API to the gRPC API.

It should be possible to create a new grpc implementation just by using the APIs in grpc-api (`io.grpc` package).

**Comment by deanorderly** [comment]
@ejona86 PERFECT!!! I 'think' that is EXACTLY what I am looking for. 

To answer your questions.  https://github.com/deanhiller/webpieces is the repo.  The tldr on what it is
1. a webserver
2. an http1.1 client
3. an http2 client
4. a bunch of pieces to construct a webserver/webclients, etc. (each piece has an api-mostly interfaces' so one can play with the legos)

I just added a gRPC/json plugin today that translates json to GRPC objects fed to a controller.  I may link to an example of that.  The next item though will be accepting gRPC binary format as well and making sure it gets to a controller/service of some sort somehow(not sure of this piece yet).

thanks for the pointers!!!

**Comment by ejona86** [comment]
Sweet! It looks like this is resolved now, so I'll go ahead and close. If you have more question (soonish) you can ask them on this issue and we can reopen, but you can also open a new issue.

I'll also point out this JSON+gRPC "example":
https://github.com/grpc/grpc-java/tree/master/examples/src/main/java/io/grpc/examples/advanced . In particular, you may find the usage of `com.google.protobuf.util.JsonFormat` to be helpful.

**Comment by deanorderly** [comment]
yup @ejona86 !!!  we are already using JsonFormat for our gRPC/json plugin which was coded up but not landed on master just yet.  Hopefully this weekend, I can figure out the binary format of it all.  thanks much!!

**Comment by deanorderly** [comment]
@ejona86 @voidzcy As a note, YEAH!! and thanks!!!  I completed the grpc-json plugin complete with no need to restart 'development server' on changes... (hot recompile on changed classes only with eclipse compiler)...
https://github.com/deanhiller/webpieces/blob/master/webserver-plugins/plugin-grpc-json/

production server of course does not even have eclipse lib nor CompilingClassLoader in it's classpath.

and we launched on one real service already.  

This weekend, I am going to dig more into your notes in creating a binary format so I can add plugin-grpc for the binary format.  THEN, one controller can server both json and binary protobuf protocol I hope.
